File: algorithm/acwing/LanQiaoBeiDayOne2023/src/Week2/day1/ACwing98/main.py
from math import sqrt
#author :袁佳林
def get(n,a):
    if n==0:
        return (0,0)
    block, leth = 1 << (2 * n - 2), 1 << (n - 1)
    (x, y), z = get(n-1, a % block), a // block
    if not z:   # z in [0, 1, 2, 3]
        return (y, x)
    elif z == 1:
        return (x, y + leth)
    elif z == 2:
        return (x + leth, y + leth)
    return (2 * leth - 1 - y, leth - 1 - x)

for _ in range(int(input())):
    n, a, b = map(int, input().split())
    x1, y1 = get(n, a-1)
    x2, y2 = get(n, b-1)
    print(f'{sqrt((x1 - x2) ** 2 + (y1 - y2) ** 2)*10:.0f}')

File: algorithm/acwing/LanQiaoBeiDayOne2023/src/Week2/day1/ACwing98/main.go
package main

/**
 * @Author 袁佳林
 * @Description
 * @Date 2023/2/28 15:15
 */
import (
    "bufio"
    "fmt"
    "math"
    "os"
)

//N const
var out = bufio.NewWriter(os.Stdout)
var in = bufio.NewReader(os.Stdin)
var t, n, a, b int
func main() {
    defer out.Flush()
    for fmt.Fscan(in, &t); t > 0; t-- {
        fmt.Fscan(in, &n, &a, &b)
        x1, y1 := calc(n, a - 1)
        x2, y2 := calc(n, b - 1)
        dx, dy := float64(x2 - x1), float64(y2 - y1)
        var res float64 = math.Sqrt(dx*dx + dy*dy)*10
        fmt.Fprintf(out, "%.0f", res)
        fmt.Fprintln(out)
    }
    fmt.Fprint(out)
}

func calc(n, m int) (int, int) {
    if n == 0 {
        return 0, 0
    }
    len, cnt := 1<<(n-1), 1<<(2*n-2)
    x, y := calc(n-1, m%cnt)
    z := m/cnt
    if z == 0 {
        return y, x
    }
    if z == 1 {
        return x, y + len
    }
    if z == 2 {
        return x + len, y + len
    }
    return 2*len-1-y, len-1-x
}
